SIP/sip-50.md

```diff
@@ -0,0 +1,174 @@
+---
+sip: 50
+title: Web Sockets API 
+description: Definition of the Web Sockets API
+author: ohager
+status: Draft
+type: standard
+category: Interface, SRC
+created: 2024-01-10
+---
+
+
+## Abstract
+This document describes the communication interface for lighweight realtime communication via WebSockets for the Node Software.
+
+## Motivation
+The Node software currently offers an extensive HTTP interface to communicate remotely with the Node software. This interface is used by software applications such as wallets. Because HTTP is a pure request-response procedure, a client application must regularly ask for updates (polling). WebSocket communication is used to establish a truly realtime connection between the client and the node software so that the server can notify connected clients directly about new blocks/transactions, for example. This allows communication to take place more efficiently according to the Hollywood principle, i.e. potentially less resource consumption (bandwidth, data volume).
+
+## Specification
+
+The new endpoint for the WebSocket connection SHALL be available under `ws[s]://<host>:<port>/events`. Once a client connects to this endpoint, it should get different real-time events from the node software.
+
+
+### Events
+
+Following events MUST be emitted:
+
+| Event             | Occurrance      | Purpose                                                 |
+|-------------------|-----------|-------------------------------------------------------------|
+| `CONNECTED`       | Immediate on connection    | Gives immediate feedback about successful connection and returns basic current state of the connected node |
+| `HEARTBEAT`       | Periodically when Idle     | Informs about a standing, healthy connection. Can be used to detect interrupted connections              |
+| `BLOCK_PUSHED`       | On each new block     | Informs that a new block was just added to the blockchain              |
+| `PENDING_TRANSACTIONS_ADDED`       | On each incoming transaction     | Informs that a transactions was submitted to the network, waiting for confirmation |
+
+
+Each event MAY have an additional payload.
+
+
+#### Anatomy
+
+The responses from the Node Software follow a unified structure. All responses MUST be serialized valid JSON (RFC8259)
+
+| Field             | Type      | Description                                                 | Example                  |
+|-------------------|-----------|-------------------------------------------------------------|--------------------------|
+| `e`               | String    | Represents the event type or action. Event name SHOULD be uppercase. | `"CONNECTED"`            |
+| `p`               | Object    | Payload containing relevant data for the event. This field is optional              |                          |
+
+#### Event `CONNECTED`
+
+This event MUST be emitted by the Node Software upon every established websocket connection.
```

Comment:
The Node Software MUST emit this event upon every established websocket connection.